File: src/bin/0017-letter-combinations-of-a-phone-number.cpp
// @leetup=info id=17 lang=cpp slug=letter-combinations-of-a-phone-number

#include <algorithm>
#include <cassert>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  std::vector<std::string> res;
  std::string s;
  std::string digits;
  std::vector<std::string> d_to_chars;

  std::vector<std::string> letterCombinations(std::string digits) {
    if (digits.empty()) {
      return {};
    }

    this->digits = digits;
    d_to_chars = {"",    "",    "abc",  "def", "ghi",
                  "jkl", "mno", "pqrs", "tuv", "wxyz"};
    assert(d_to_chars.size() == 10);

    dfs(0);
    return res;
  }

  void dfs(size_t i) {
    if (i == digits.size()) {
      res.push_back(s);
      return;
    }
    for (char c : d_to_chars[digits[i] - '0']) {
      s.push_back(c);
      dfs(i + 1);
      s.pop_back();
    }
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0017-letter-combinations-of-a-phone-number.py
# @leetup=info id=17 lang=python3 slug=letter-combinations-of-a-phone-number



# @leetup=code


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        m = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }

        res = []
        letters = []

        def dfs(i: int):
            if i == len(digits):
                res.append("".join(letters))
                return
            for c in m[digits[i]]:
                letters.append(c)
                dfs(i + 1)
                letters.pop()

        dfs(0)

        return res


# @leetup=code
